### PR TITLE
Add k6

### PR DESCRIPTION
**Pull request by patrick-tavares**
Modeled after Stormforge-perf with appropriate changes specific to k6. @jgustie or @tperol please review.

### CODE REVIEW

File: k6/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM grafana/k6
+
+USER root
+RUN apk add --no-cache \
+		curl \
+		jq
```
Comment (jdplatt) [inline k6/Dockerfile:6]: I don't think you need jq anymore

Comment (patrick-tavares) [reply k6/Dockerfile]: Removed.

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+**/.DS_Store
```
Comment (jgustie) [inline .gitignore:1]: This should be in your global Git ignore file, not committed in the repo.

File: k6/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
+FROM grafana/k6
+
+USER root
+RUN apk add --no-cache \
+		curl
+COPY docker-entrypoint.sh /usr/local/bin/
+COPY handleSummary.js .
+RUN chmod +x /usr/local/bin/docker-entrypoint.sh
+RUN chown 12345 /usr/local/bin/docker-entrypoint.sh
```
Comment (jgustie) [inline k6/Dockerfile:9]: Set the permissions appropriately on the file in Git, you should not need to change either the file permissions or owner in the Dockerfile.

File: k6/docker-entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+if ! [ -f "${TEST_CASE_FILE}" ]; then
+        echo "A test case JS file must be provided." 1>&2
+        exit 1
+fi
+
+# Check that we have a Prometheus Pushgateway defined
+if [ -z "${PUSHGATEWAY_URL}" ]; then
+        echo "No Pushgateway URL" 1>&2
+				exit 1
```
Comment (jgustie) [inline k6/docker-entrypoint.sh:17]: Make sure to use tabs for indentation in shell scripts.

File: k6/docker-entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+fi
+
+COMBINED_TEST_CASE_FILE="/home/k6/combined_load_script.js"
```
Comment (jgustie) [inline k6/docker-entrypoint.sh:20]: You shouldn't be writing to a location that isn't documented by a volume mount, i.e. this should be in `/tmp`.

That being said, you can probably do away with the need for the file system file altogether:
```
k6 run <(grep -q "export \+function \+handleSummary" file.js && cat file.js || cat file.js handleSummary.js)
```

File: k6/docker-entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+k6 run "${COMBINED_TEST_CASE_FILE}"
+
+# If
+if ! [ -f "${TEST_CASE_OUTPUT}" ]; then
+        echo "Test case output was not found. Unable to push to Prometheus." 1>&2
+        exit 1
+fi
+
+# Push to the Prometheus PushGateway
+cat ${TEST_CASE_OUTPUT} | curl --data-binary @- "${PUSHGATEWAY_URL}"
```
Comment (jgustie) [inline k6/docker-entrypoint.sh:37]: This should be combined with the `-n "${PUSHGATEWAY_URL}"` to make sure there is a URL to push to.

File: k6/output.json [hunk 7]
```diff
@@ -0,0 +1,114 @@
+{
```
Comment (jgustie) [inline k6/output.json:1]: Is this file necessary?

**Review by jgustie — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (patrick-tavares) [reply .gitignore:1]: Ugh, I staged all files in a hurry to get this PR in and forgot to filter this out. I am embarrassed :)

File: k6/output.json (hunk 7 above)
Comment (patrick-tavares) [reply k6/output.json:1]: I am embarrassed here too :)

File: k6/docker-entrypoint.sh (hunk 6 above)
Comment (patrick-tavares) [reply k6/docker-entrypoint.sh:37]: I tested for PUSHGATEWAY_URL lines 15 - 18 to exit right away on failure (and skip all further operations). Would you still prefer to wrap the PUSHGATEWAY_URL use attempt with a check?

File: k6/Dockerfile (hunk 3 above)
Comment (patrick-tavares) [reply k6/Dockerfile]: Changes made as specified.

File: k6/docker-entrypoint.sh (hunk 4 above)
Comment (patrick-tavares) [reply k6/docker-entrypoint.sh]: Changes made as specified.

File: k6/docker-entrypoint.sh (hunk 5 above)
Comment (patrick-tavares) [reply k6/docker-entrypoint.sh]: Re-worked to use /tmp to stage the combined test case as k6 cannot run using an in-lined js file.

File: k6/docker-entrypoint.sh (hunk 6 above)
Comment (jgustie) [reply k6/docker-entrypoint.sh]: You should eliminate the check on 15-18 in favor of only sending the output if the push gateway URL is defined (i.e. the test didn't fail if couldn't report results; the _trial_ will still fail because it won't be able to collect those results later).

Comment (patrick-tavares) [reply k6/docker-entrypoint.sh]: Given that a load test could take many minutes to run, wouldn't it make sense to fail on no PUSHGATEWAY_URL before the load test begins so that the trial can fail faster and thus be quicker to debug?

Comment (patrick-tavares) [reply k6/docker-entrypoint.sh]: @jgustie can we resolve this?

Comment (jgustie) [reply k6/docker-entrypoint.sh]: The other trial jobs will run without `PUSHGATEWAY_URL` set, this should be no different.

Comment (patrick-tavares) [reply k6/docker-entrypoint.sh]: Reworked check for PUSHGATEWAY_URL to check for successful test metric push if and only if PUSHGATEWAY_URL is not empty.

File: k6/README.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    import http from 'k6/http';
+    import { sleep } from 'k6';
+
+    export default function () {
+      http.get('https://test.k6.io');
+      sleep(10);
+    }
+```
+
+NOTE: If you are using this in an experiment, keep in mind that some values are set automatically. In particular, the `backoffLimit`, `restartPolicy`, and `PUSHGATEWAY_URL` environment variable are all introduced when evaluating a trial's job template.
```
Comment (zeisss) [inline k6/README.md:157]: The PUSHGATEWAY_URL is only added, if the `prometheus` setupTask is used.

File: k6/docker-entrypoint.sh (hunk 6 above)
Comment (patrick-tavares) [reply k6/docker-entrypoint.sh]: @jgustie are we good here?

**Comment by zeisss** [comment]
@patrick-tavares  @ztreinhart and I are started testing this branch / using it in some experiments. we will make some comments here if we have some feedback.

File: k6/docker-entrypoint.sh [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  QUERY_BASE=$(echo $PUSHGATEWAY_URL | cut -d\/ -f1-3)
+  TEST_VALUE=$(curl -s "${QUERY_BASE}/api/v1/metrics" | jq -r '.data[0].test_metric.metrics[0].value')
+  printf "Queried test mertric value: \"${TEST_VALUE}\" ..."
+  if  [ $TEST_VALUE -ne 42 ]; then
+    echo " FAILED!"
+    echo "\tTest metric POST was successful but test metric verification query failed. Exiting. "
+    return 1
+  fi
+  echo " PASSED!"
+}
```
Comment (jgustie) [inline k6/docker-entrypoint.sh:36]: I think I would have just gone with something like `curl --fail --request-target '/-/ready' $PUSHGATEWAY_URL` to test if the Push Gateway was ready, this seems excessive. I'm not going to hold up merging, but I'm also not going to say we aren't going to change it in the future to "fix" the issue of posting dummy metrics.

File: k6/README.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+as shown below. If your test case already defines a handleSummary() function it
+will not be overridden and, consequently, you will need to ensure your output
+format can be pushed to Prometheus and you appropriately reference them in your
+experiment files.
+
+## Configuration
+
+| Environment Variable | Description |
+| -------------------- | ----------- |
+| `TEST_CASE_FILE`     | Path to the test case file mounted in the container. Defaults to /scrips/load_script.js. Associated VolumeMount required that contains this file. If the file is not found, the trial fails. |
```
Comment (ztreinhart) [inline k6/README.md:23]:
```suggestion
| `TEST_CASE_FILE`     | Path to the test case file mounted in the container. Defaults to /scripts/load_script.js. Associated VolumeMount required that contains this file. If the file is not found, the trial fails. |
```

File: k6/README.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+provided as either data from a ConfigMap
+(mounted in as a Volume) or as an existing file on a PV mounted in as a Volume.
+In both cases, appropriately specified environment variables below are required.
+
+In order to make the k6 output metrics usable by the Optimize Pro ML engine
+the metrics need to be massaged to be pushable to Prometheus. As a result, we
+make use of k6's handleSummary() override function to format the output metrics
+as shown below. If your test case already defines a handleSummary() function it
+will not be overridden and, consequently, you will need to ensure your output
+format can be pushed to Prometheus and you appropriately reference them in your
```
Comment (ztreinhart) [inline k6/README.md:16]:
```suggestion
format can be pushed to Prometheus as metrics and you appropriately reference them in your
```

**Review by ztreinhart — COMMENTED** [review]
Just a few small things I noticed in the readme.

**Comment by ztreinhart** [comment]
@zeisss and I have built and played with this trial pod as part of our starting work toward supporting more 3rd party performance testing tools, and it seems to work pretty well. 

We have a simple experiment using this pod in a PR in [the examples repo](https://github.com/thestormforge/examples/tree/stormforge-testapp/stormforger-testapp/k6). Feel free to link to it or steal from it to enhance your readme.

File: k6/README.md (hunk 8 above)
Comment (zeisss) [reply k6/README.md:157]:
```suggestion
NOTE: If you are using this in an experiment, keep in mind that some values are set automatically. In particular, the `backoffLimit`, `restartPolicy`, and `PUSHGATEWAY_URL` environment variable are all introduced when evaluating a trial's job template. `PUSHGATEWAY_URL` requires the usage of the `prometheus` setupTask though.
```

File: .github/workflows/main.yaml [hunk 12]
```diff
@@ -72,3 +72,27 @@ jobs:
         labels: ${{ steps.locust_docker_meta.outputs.labels }}
     - name: Locust image digest
       run: echo ${{ steps.locust_docker_build.outputs.digest }}
+
+    - name: Set up Docker Buildx for K6
+      uses: docker/setup-buildx-action@v1
+      id: k6_builder
+    - name: K6 Docker meta
+      id: k6_docker_meta
+      uses: crazy-max/ghaction-docker-meta@v1
```
Comment (tisba) [inline .github/workflows/main.yaml:81]: https://github.com/crazy-max/ghaction-docker-buildx

> This action is ARCHIVED and will not receive any updates, update your workflows to use the official Docker actions.

Comment (jgustie) [reply .github/workflows/main.yaml:81]: See #9

**Review by tisba — APPROVED** [review]

File: .github/workflows/test-k6.yaml [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+          -v ${PWD}/test:/test \
+          --env PUSHGATEWAY_URL=http://pushgateway:9091/metrics/job/sf_k6_test \
+          --env TEST_CASE_FILE=/test/simple.js \
+          --env SF_K6_TARGET=http://testapp:8080 \
+          sf-k6-test
+
+    - name: Test metrics from Pushgateway
+      working-directory: k6
+      run: |
+        echo "Scarping metrics from Pushgateway..."
```
Comment (ztreinhart) [inline .github/workflows/test-k6.yaml:59]:
```suggestion
        echo "Scraping metrics from Pushgateway..."
```

File: k6/docker-entrypoint.sh [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+TEST_CASE_FILE=${TEST_CASE_FILE:-"/scripts/script.js"}
+
+if ! [ -f "${TEST_CASE_FILE}" ]; then
+  echo "A test case JS file must be provided."
+  exit 1
+fi
+
+# NOTE: handleSummary.js currently hardcodes the output file, we're instructing
+#       k6 to generate when its done.
+PROMETHEUS_OUTPUT_FILE="prometheus.txt"
```
Comment (ztreinhart) [inline k6/docker-entrypoint.sh:15]:
```suggestion
PROMETHEUS_OUTPUT_FILE="/tmp/prometheus.txt"
```

**Review by ztreinhart — DISMISSED** [review]

**Review by ztreinhart — APPROVED** [review]
